crates/sequencer/src/serde.rs

```diff
@@ -0,0 +1,106 @@
+use std::{io, path::Path};
+
+use crate::state::State;
+use blockifier::{
+    execution::contract_class::ContractClass, state::cached_state::ContractStorageKey,
+};
+use rustc_hash::FxHashMap;
+use serde::{Deserialize, Serialize};
+use starknet_api::{
+    core::{ClassHash, CompiledClassHash, ContractAddress, Nonce},
+    hash::StarkFelt,
+};
+
+use thiserror::Error;
+
+pub trait DumpLoad {
+    fn dump_state_to_file(self, file_path: &Path) -> Result<(), SerializationError>;
+
+    fn load_state_from_file(file_path: &Path) -> Result<Self, SerializationError>
+    where
+        Self: Sized;
+}
+
+#[derive(Error, Debug)]
+pub enum SerializationError {
+    #[error(transparent)]
+    IoError(#[from] io::Error),
+    #[error(transparent)]
+    SerdeJsonError(#[from] serde_json::Error),
+}
+
+#[derive(Clone, Debug, Serialize, Deserialize, Default)]
+pub struct SerializableState {
+    pub classes: FxHashMap<ClassHash, ContractClass>,
+    pub compiled_classes_hash: FxHashMap<ClassHash, CompiledClassHash>,
+    pub contracts: FxHashMap<ContractAddress, ClassHash>,
+    #[serde(with = "serialize_contract_storage")]
+    pub storage: FxHashMap<ContractStorageKey, StarkFelt>,
+    pub nonces: FxHashMap<ContractAddress, Nonce>,
+}
+
+impl From<State> for SerializableState {
+    fn from(state: State) -> Self {
+        Self {
+            classes: state.classes,
+            compiled_classes_hash: state.compiled_class_hashes,
+            contracts: state.contracts,
+            storage: state.storage,
+            nonces: state.nonces,
+        }
+    }
+}
+
+mod serialize_contract_storage {
+    use blockifier::state::cached_state::ContractStorageKey;
+    use rustc_hash::FxHashMap;
+    use serde::de::{Deserialize, Deserializer};
+    use serde::ser::{Serialize, Serializer};
+    use starknet_api::hash::StarkFelt;
+    use std::collections::HashMap;
+
+    pub fn serialize<S>(
+        map: &FxHashMap<ContractStorageKey, StarkFelt>,
+        serializer: S,
+    ) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        let mut transformed: HashMap<String, String> = HashMap::new();
+        for (contract_storage_key, storage_value) in map.iter() {
+            let key_str = serde_json::to_string(&contract_storage_key).unwrap();
+            let value_str = serde_json::to_string(&storage_value).unwrap();
+            transformed.insert(key_str, value_str);
+        }
```

Comment:
Mmmmm do we need to make the intermediate mapping? I feel like we could just serialize each (k,v) pair straight into the serializer, skipping this temp value.
Found this which could be of use
https://stackoverflow.com/questions/51276896/how-do-i-use-serde-to-serialize-a-hashmap-with-structs-as-keys-to-json
I think there should be an easier way to do this where we also don't need to implement our own deser method.

Reply:
Basically the issue is that you need the key to be a string, but then when you put it into a map and serialize this, what you get is actually a serialize string. Then when serde tries to deserialize the key into a tuple, it fails because the key is now a serialized string. I think we can improve this.

Reply:
reduced the complexity via [this](https://github.com/kkrt-labs/ef-tests/pull/555/commits/3d27d5b9aa81e1ce48db4c12aba286d6eb362ff7) commit.